<a href="https://colab.research.google.com/github/ashwin-yedte/visual-intelligence-travel-finance/blob/main/VL_Encoding_Framework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**VL ENCODING FRAMEWORK** - WITH COMPREHENSIVE IMAGE PREPROCESSING
Visual-Language Encoding Infrastructure for Indian Travel Destinations




COMPLETE FEATURE SET:
1.  CLIP tensor extraction bug
2. Comprehensive image preprocessing (EXIF, RGB, aspect ratio)
3. Image validation and quality checks
4. Prompt validation (normalization, token limits)
5. Quality reports and statistics
6. Error recovery and logging

PREPROCESSING PIPELINES:
- IMAGE: EXIF orientation → RGB conversion → Aspect ratio → Padding → Validation
- PROMPT: Token validation → Duplicate detection → Quality metrics

OBJECTIVES:
1. Pre-compute CLIP embeddings for all landmark images
2. Extract semantic prompts using 2,200 CLIP prompt library
3. Create searchable database for similarity matching
4. Enable two-stage matching (visual + semantic)


WORKFLOW:
- Stage 1: Pre-compute prompt embeddings (one-time, 2,200 prompts)
- Stage 2: Process each image (extract CLIP embedding + prompts)
- Stage 3: Aggregate per destination
- Stage 4: Create search indices

================================================================

INSTALL PACKAGES

================================================================

In [1]:
!pip install -q transformers torch pillow tqdm
import warnings
warnings.filterwarnings('ignore')

print("="*80)
print("PACKAGES INSTALLED")
print("="*80)

PACKAGES INSTALLED


================================================================

IMPORT PACKAGES

================================================================

In [2]:
from google.colab import drive
import os
import json
import numpy as np
from PIL import Image, ImageOps
import torch
from transformers import CLIPModel, CLIPProcessor
from tqdm import tqdm
from datetime import datetime
import pickle
import warnings
from typing import Tuple, Dict, Any, List

warnings.filterwarnings('ignore')



# ================================================================
   SETUP AND MOUNT Google Drive   
# ================================================================


In [3]:
print("="*80)
print("VISUAL LANGUAGE ENCODING FRAMEWORK")
print("="*80)

drive.mount('/content/drive')

BASE_PATH = '/content/drive/MyDrive/visual-intelligence-travel-finance'

LANDMARKS_PATH = f'{BASE_PATH}/data/landmarks'
METADATA_PATH = f'{LANDMARKS_PATH}/metadata.json'
PROMPT_LIBRARY_PATH = f'{BASE_PATH}/data/prompt_library/clip_prompts_india_themes_semantic.json'

VL_ENCODING_PATH = f'{BASE_PATH}/data/vl_encoding'
EMBEDDINGS_PATH = f'{VL_ENCODING_PATH}/embeddings'
PROMPTS_PATH = f'{VL_ENCODING_PATH}/prompts'
REPORTS_PATH = f'{VL_ENCODING_PATH}/reports'

os.makedirs(EMBEDDINGS_PATH, exist_ok=True)
os.makedirs(f'{PROMPTS_PATH}/image_prompts', exist_ok=True)
os.makedirs(REPORTS_PATH, exist_ok=True)

print(f"Directories created")
print("="*80)




VISUAL LANGUAGE ENCODING FRAMEWORK
Mounted at /content/drive
Directories created


================================================================

STEP 1: PROMPT VALIDATION

================================================================

    
    Validates and analyzes text prompts for CLIP processing.
    Features:
    - Token length validation (CLIP limit: 77 tokens)    
    - Quality metrics and reporting
    

In [4]:
print("\n" + "="*80)
print("PROMPT VALIDATION MODULE")
print("="*80)

class PromptValidator:
    def __init__(self, processor: CLIPProcessor):
        """
        Initialize validator with CLIP processor.

        Args:
            processor: CLIPProcessor for tokenization
        """
        self.processor = processor
        self.validation_log = []
        print("PromptValidator initialized")

    def validate_prompt(self, prompt: str, prompt_id: str = None) -> Dict[str, Any]:
        """
        Validate a single prompt.

        Args:
            prompt: Text prompt to validate
            prompt_id: Optional identifier for logging

        Returns:
            Dictionary with validation results
        """
        issues = []
        warnings = []

        if not prompt or not prompt.strip():
            issues.append("Empty prompt")
            return {
                'valid': False,
                'issues': issues,
                'warnings': warnings,
                'prompt': prompt,
                'prompt_id': prompt_id
            }

        prompt_clean = prompt.strip()

        char_length = len(prompt_clean)
        if char_length < 5:
            warnings.append(f"Very short prompt ({char_length} chars)")
        if char_length > 150:
            warnings.append(f"Long prompt ({char_length} chars)")

        try:
            tokens = self.processor.tokenizer(prompt_clean, truncation=False)
            token_count = len(tokens['input_ids'])

            if token_count > 77:
                issues.append(f"Exceeds CLIP token limit: {token_count}/77 tokens")
            elif token_count > 60:
                warnings.append(f"Near token limit: {token_count}/77 tokens")

        except Exception as e:
            issues.append(f"Tokenization failed: {str(e)}")
            token_count = -1

        problem_chars = ['@', '#', '$', '%', '^', '*', '|', '\\']
        found_chars = [c for c in problem_chars if c in prompt_clean]
        if found_chars:
            warnings.append(f"Contains special characters: {found_chars}")

        result = {
            'valid': len(issues) == 0,
            'issues': issues,
            'warnings': warnings,
            'prompt': prompt_clean,
            'prompt_id': prompt_id,
            'char_length': char_length,
            'token_count': token_count if token_count != -1 else None
        }

        self.validation_log.append(result)
        return result

    def validate_prompt_library(self, prompt_library: Dict) -> Dict[str, Any]:
        """
        Validate entire prompt library.

        Args:
            prompt_library: Dictionary of themes -> categories -> prompts

        Returns:
            Validation report with statistics
        """
        print("Validating prompt library...")

        report = {
            'total_prompts': 0,
            'valid_prompts': 0,
            'prompts_with_issues': 0,
            'prompts_with_warnings': 0,
            'issues_found': [],
            'warnings_found': [],
            'token_stats': {
                'min': float('inf'),
                'max': 0,
                'mean': 0,
                'over_limit': 0
            }
        }

        token_counts = []

        for theme, categories in prompt_library.items():
            for category, prompts in categories.items():
                for idx, prompt in enumerate(prompts):
                    prompt_id = f"{theme}/{category}/{idx}"

                    result = self.validate_prompt(prompt, prompt_id)
                    report['total_prompts'] += 1

                    if result['valid']:
                        report['valid_prompts'] += 1
                    else:
                        report['prompts_with_issues'] += 1
                        report['issues_found'].extend([
                            {'prompt_id': prompt_id, 'issue': issue}
                            for issue in result['issues']
                        ])

                    if result['warnings']:
                        report['prompts_with_warnings'] += 1
                        report['warnings_found'].extend([
                            {'prompt_id': prompt_id, 'warning': warning}
                            for warning in result['warnings']
                        ])

                    if result['token_count'] is not None:
                        token_counts.append(result['token_count'])
                        if result['token_count'] > 77:
                            report['token_stats']['over_limit'] += 1

        if token_counts:
            report['token_stats']['min'] = int(min(token_counts))
            report['token_stats']['max'] = int(max(token_counts))
            report['token_stats']['mean'] = float(np.mean(token_counts))

        return report

    def generate_report(self, validation_results: Dict, output_file: str = None) -> str:
        """
        Generate human-readable validation report.

        Args:
            validation_results: Results from validate_prompt_library()
            output_file: Optional path to save report

        Returns:
            Report text
        """
        report_lines = []
        report_lines.append("=" * 80)
        report_lines.append("PROMPT LIBRARY VALIDATION REPORT")
        report_lines.append("=" * 80)
        report_lines.append("")

        report_lines.append("SUMMARY")
        report_lines.append("-" * 80)
        report_lines.append(f"Total prompts: {validation_results['total_prompts']}")
        report_lines.append(f"Valid prompts: {validation_results['valid_prompts']}")
        report_lines.append(f"Prompts with issues: {validation_results['prompts_with_issues']}")
        report_lines.append(f"Prompts with warnings: {validation_results['prompts_with_warnings']}")
        report_lines.append("")

        report_lines.append("TOKEN STATISTICS")
        report_lines.append("-" * 80)
        stats = validation_results['token_stats']
        report_lines.append(f"Min tokens: {stats['min']}")
        report_lines.append(f"Max tokens: {stats['max']}")
        report_lines.append(f"Mean tokens: {stats['mean']:.1f}")
        report_lines.append(f"Prompts over limit (77): {stats['over_limit']}")
        report_lines.append("")

        if validation_results['issues_found']:
            report_lines.append("CRITICAL ISSUES")
            report_lines.append("-" * 80)
            for item in validation_results['issues_found']:
                report_lines.append(f"  [{item['prompt_id']}] {item['issue']}")
            report_lines.append("")
        else:
            report_lines.append("No critical issues found")
            report_lines.append("")

        if validation_results['warnings_found']:
            report_lines.append("WARNINGS")
            report_lines.append("-" * 80)
            for item in validation_results['warnings_found'][:10]:
                report_lines.append(f"  [{item['prompt_id']}] {item['warning']}")
            if len(validation_results['warnings_found']) > 10:
                report_lines.append(f"  ... and {len(validation_results['warnings_found']) - 10} more")
            report_lines.append("")

        report_lines.append("=" * 80)

        report_text = "\n".join(report_lines)

        if output_file:
            with open(output_file, 'w') as f:
                f.write(report_text)
            print(f"Report saved to: {output_file}")

        return report_text


print("PromptValidator class ready")
print("="*80)


PROMPT VALIDATION MODULE
PromptValidator class ready


================================================================

STEP 2: IMAGE PREPROCESSING

================================================================

    Comprehensive image preprocessing for CLIP model.
    
    Handles:
    - EXIF orientation correction
    - RGB conversion from any color mode
    - Aspect ratio preservation
    - High-quality resampling
    - White padding for consistent dimensions


In [5]:
print("\n" + "="*80)
print("IMAGE PREPROCESSING")
print("="*80)

class ImagePreprocessor:
    """Comprehensive image preprocessing for CLIP."""

    def __init__(self, target_size: Tuple[int, int] = (224, 224)):
        self.target_size = target_size
        print(f"ImagePreprocessor initialized (target: {target_size})")

    def preprocess_image(self, image_path: str) -> Image.Image:
        """Load and preprocess image."""
        try:
            img = Image.open(image_path)
            img = self._fix_orientation(img)

            if img.mode != 'RGB':
                if img.mode == 'RGBA':
                    background = Image.new('RGB', img.size, (255, 255, 255))
                    background.paste(img, mask=img.split()[3])
                    img = background
                else:
                    img = img.convert('RGB')

            img = self._resize_with_padding(img, self.target_size)
            return img

        except Exception as e:
            raise Exception(f"Failed to preprocess {image_path}: {str(e)}")

    def _fix_orientation(self, img: Image.Image) -> Image.Image:
        """Fix EXIF orientation."""
        try:
            img = ImageOps.exif_transpose(img)
        except:
            pass
        return img

    def _resize_with_padding(self, img: Image.Image, target_size: Tuple[int, int]) -> Image.Image:
        """Resize with aspect ratio preservation."""
        img_ratio = img.width / img.height
        target_ratio = target_size[0] / target_size[1]

        if img_ratio > target_ratio:
            new_width = target_size[0]
            new_height = int(new_width / img_ratio)
        else:
            new_height = target_size[1]
            new_width = int(new_height * img_ratio)

        img = img.resize((new_width, new_height), Image.Resampling.LANCZOS)

        canvas = Image.new('RGB', target_size, (255, 255, 255))
        offset_x = (target_size[0] - new_width) // 2
        offset_y = (target_size[1] - new_height) // 2
        canvas.paste(img, (offset_x, offset_y))

        return canvas

    def validate_image(self, image_path: str, max_size_mb: float = 10.0) -> Dict[str, Any]:
        """Validate image."""
        try:
            if not os.path.exists(image_path):
                return {'valid': False, 'error': 'File not found'}

            size_mb = os.path.getsize(image_path) / (1024 * 1024)
            if size_mb > max_size_mb:
                return {'valid': False, 'error': f'File too large: {size_mb:.2f}MB'}

            img = Image.open(image_path)
            img_format = img.format
            dimensions = img.size
            img.close()

            return {'valid': True, 'size_mb': size_mb, 'format': img_format, 'dimensions': dimensions}

        except Exception as e:
            return {'valid': False, 'error': str(e)}


preprocessor = ImagePreprocessor(target_size=(224, 224))
print("="*80)


IMAGE PREPROCESSING
ImagePreprocessor initialized (target: (224, 224))


================================================================

STEP 3: LOAD CLIP MODEL

================================================================

In [6]:
print("\n" + "="*80)
print("LOADING CLIP MODEL")
print("="*80)

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Device: {device}")

model_name = "openai/clip-vit-base-patch32"
print(f"Model: {model_name}")
print("Loading...")

model = CLIPModel.from_pretrained(model_name)
processor = CLIPProcessor.from_pretrained(model_name)

model.to(device)
model.eval()

sample_text = ["test"]
inputs = processor(text=sample_text, return_tensors="pt", padding=True)
inputs = {k: v.to(device) for k, v in inputs.items()}
with torch.no_grad():
    test_output = model.get_text_features(**inputs)
    if torch.is_tensor(test_output):
        embedding_dim = test_output.shape[-1]
    else:
        embedding_dim = 512

print(f"\nModel loaded")
print(f"  Parameters: {sum(p.numel() for p in model.parameters())/1e6:.1f}M")
print(f"  Embedding dimension: {embedding_dim}")
if device == "cuda":
    print(f"  GPU: {torch.cuda.get_device_name(0)}")

print("="*80)


LOADING CLIP MODEL
Device: cpu
Model: openai/clip-vit-base-patch32
Loading...


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/605M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/605M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

CLIPModel LOAD REPORT from: openai/clip-vit-base-patch32
Key                                  | Status     |  | 
-------------------------------------+------------+--+-
text_model.embeddings.position_ids   | UNEXPECTED |  | 
vision_model.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


preprocessor_config.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

The image processor of type `CLIPImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. 


tokenizer_config.json:   0%|          | 0.00/592 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/389 [00:00<?, ?B/s]


Model loaded
  Parameters: 151.3M
  Embedding dimension: 512


================================================================

STEP 4: LOAD IMAGE META DATA  AND PROMPT LIBRARY

================================================================

In [7]:
print("\n" + "="*80)
print("LOADING DATA")
print("="*80)

with open(METADATA_PATH, 'r') as f:
    metadata = json.load(f)

print(f"Metadata: {metadata['total_images']} images, {metadata['total_destinations']} destinations")

# ADD THIS: Initialize pipeline_status if it doesn't exist
if 'pipeline_status' not in metadata:
    metadata['pipeline_status'] = {
        'embeddings_computed': False,
        'prompts_extracted': False,
        'prompts_validated': False
    }
    print("Initialized pipeline_status in metadata")

with open(PROMPT_LIBRARY_PATH, 'r') as f:
    prompt_library = json.load(f)

total_prompts = sum(sum(len(prompts) for prompts in categories.values())
                    for categories in prompt_library.values())
print(f"Prompt library: {total_prompts} prompts")

print("="*80)


LOADING DATA
Metadata: 805 images, 168 destinations
Initialized pipeline_status in metadata
Prompt library: 2200 prompts


================================================================

STEP 5: VALIDATE PROMPT LIBRARY

================================================================

In [8]:
print("\n" + "="*80)
print("VALIDATING PROMPT LIBRARY")
print("="*80)

validator = PromptValidator(processor)
validation_results = validator.validate_prompt_library(prompt_library)

validation_report = validator.generate_report(
    validation_results,
    output_file=f'{REPORTS_PATH}/prompt_validation_report.txt'
)

print("\n" + validation_report)

if validation_results['prompts_with_issues'] > 0:
    print(f"\nWARNING: {validation_results['prompts_with_issues']} prompts have critical issues!")
    print("Review the validation report before proceeding.")
else:
    print("\nAll prompts passed validation!")

print("="*80)



VALIDATING PROMPT LIBRARY
PromptValidator initialized
Validating prompt library...
Report saved to: /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/reports/prompt_validation_report.txt

PROMPT LIBRARY VALIDATION REPORT

SUMMARY
--------------------------------------------------------------------------------
Total prompts: 2200
Valid prompts: 2200
Prompts with issues: 0
Prompts with warnings: 0

TOKEN STATISTICS
--------------------------------------------------------------------------------
Min tokens: 17
Max tokens: 28
Mean tokens: 21.3
Prompts over limit (77): 0

No critical issues found


All prompts passed validation!


================================================================

EMBEDDING EXTRACTION FUNCTIONS

================================================================

In [9]:
print("\n" + "="*80)
print("DEFINING EXTRACTION FUNCTIONS")
print("="*80)

def extract_clip_features(outputs):
    """Universal fix for extracting tensors from CLIP outputs."""
    # PRIORITY 1: Direct tensor
    if torch.is_tensor(outputs):
        return outputs

    # PRIORITY 2: pooler_output (CLIP's projected features)
    if hasattr(outputs, 'pooler_output') and outputs.pooler_output is not None:
        return outputs.pooler_output

    # PRIORITY 3: text_embeds or image_embeds (newer CLIP versions)
    if hasattr(outputs, 'text_embeds') and outputs.text_embeds is not None:
        return outputs.text_embeds
    if hasattr(outputs, 'image_embeds') and outputs.image_embeds is not None:
        return outputs.image_embeds

    # PRIORITY 4: last_hidden_state (take CLS token)
    if hasattr(outputs, 'last_hidden_state') and outputs.last_hidden_state is not None:
        return outputs.last_hidden_state[:, 0, :]

    # PRIORITY 5: Try indexing
    try:
        return outputs[0]
    except:
        pass

    # PRIORITY 6: Try converting to tensor
    try:
        return torch.tensor(outputs)
    except:
        pass

    raise ValueError(f"Could not extract 512-dim projected CLIP features from model output type: {type(outputs)}. Object: {outputs}")



def extract_image_embedding(image_path, model, processor, preprocessor, device):
    """Extract CLIP embedding from image with preprocessing."""
    try:
        validation = preprocessor.validate_image(image_path)
        if not validation['valid']:
            print(f"Skipping {image_path}: {validation['error']}")
            return None

        img = preprocessor.preprocess_image(image_path)

        inputs = processor(images=img, return_tensors="pt", padding=True)
        inputs = {k: v.to(device) for k, v in inputs.items()}

        with torch.no_grad():
            outputs = model.get_image_features(**inputs)
            image_features = extract_clip_features(outputs)
            image_features = image_features / image_features.norm(dim=-1, keepdim=True)

        return image_features.cpu().numpy()[0]

    except Exception as e:
        print(f"\nERROR in extract_image_embedding for {image_path}: {e}")
        return None


def extract_text_embeddings_batch(texts, model, processor, device):
    """Extract CLIP embeddings for batch of text prompts."""
    try:
        inputs = processor(text=texts, return_tensors="pt", padding=True, truncation=True)
        inputs = {k: v.to(device) for k, v in inputs.items()}

        with torch.no_grad():
            outputs = model.get_text_features(**inputs)
            text_features = extract_clip_features(outputs)
            text_features = text_features / text_features.norm(dim=-1, keepdim=True)

        return text_features.cpu().numpy()

    except Exception as e:
        print(f"\nERROR in extract_text_embeddings_batch: {e}")
        return None


print("Extraction functions defined")
print("="*80)



DEFINING EXTRACTION FUNCTIONS
Extraction functions defined


================================================================

STEP 6: PRE-COMPUTE PROMPT EMBEDDINGS


    Pre-compute embeddings for all 2,200 prompts
    
    FIXED: Properly extract tensor from CLIP output object
    
    Returns:
        Dictionary with structure:
        {
            'Beach': {
                'LandscapeType': [
                    {'text': 'prompt text', 'embedding': np.array(512,)},
                    ...
                ],
                ...
            },
            ...
        }


================================================================

In [10]:
print("\n" + "="*80)
print("STAGE 6: PRE-COMPUTING PROMPT EMBEDDINGS")
print("="*80)

def precompute_prompt_embeddings(prompt_library, model, processor, device):
    """Pre-compute embeddings for all prompts."""

    prompt_embeddings = {}
    total_prompts = sum(sum(len(p) for p in cats.values()) for cats in prompt_library.values())

    print(f"Total prompts: {total_prompts}")

    with tqdm(total=total_prompts, desc="Computing prompt embeddings") as pbar:
        for theme, categories in prompt_library.items():
            prompt_embeddings[theme] = {}

            for category, prompts in categories.items():
                prompt_embeddings[theme][category] = []

                batch_size = 32
                for i in range(0, len(prompts), batch_size):
                    batch = prompts[i:i+batch_size]

                    embeddings = extract_text_embeddings_batch(batch, model, processor, device)

                    if embeddings is not None:
                        for j, prompt_text in enumerate(batch):
                            prompt_embeddings[theme][category].append({
                                'text': prompt_text,
                                'embedding': embeddings[j]
                            })

                    pbar.update(len(batch))

    return prompt_embeddings


prompt_embeddings_file = f'{EMBEDDINGS_PATH}/prompt_embeddings.pkl'

# Force re-computation to ensure consistency after fix
force_recompute_prompts = True

if os.path.exists(prompt_embeddings_file) and not force_recompute_prompts:
    print("Found cached prompt embeddings...")

    with open(prompt_embeddings_file, 'rb') as f:
        cached_embeddings = pickle.load(f)

    for theme, cats in cached_embeddings.items():
        for cat, prompts in cats.items():
            if prompts:
                cached_dim = prompts[0]['embedding'].shape[0]
                break
        break

    print(f"  Cached dimension: {cached_dim}")
    print(f"  Current model dimension: {embedding_dim}")

    if cached_dim == embedding_dim:
        print("  Dimensions match - using cache")
        prompt_embeddings = cached_embeddings
    else:
        print("  Dimension mismatch - recomputing...")
        os.remove(prompt_embeddings_file)
        prompt_embeddings = precompute_prompt_embeddings(prompt_library, model, processor, device)

        with open(prompt_embeddings_file, 'wb') as f:
            pickle.dump(prompt_embeddings, f)
        print(f"  Saved new embeddings")
else:
    if os.path.exists(prompt_embeddings_file) and force_recompute_prompts:
        print("Forcing recomputation of prompt embeddings, ignoring cached file.")
        os.remove(prompt_embeddings_file) # Ensure old cache is removed
    else:
        print("Computing prompt embeddings...")

    prompt_embeddings = precompute_prompt_embeddings(prompt_library, model, processor, device)

    with open(prompt_embeddings_file, 'wb') as f:
        pickle.dump(prompt_embeddings, f)
    print(f"Saved to: {prompt_embeddings_file}")

print("="*80)



STAGE 6: PRE-COMPUTING PROMPT EMBEDDINGS
Computing prompt embeddings...
Total prompts: 2200


Computing prompt embeddings: 100%|██████████| 2200/2200 [01:31<00:00, 24.00it/s]

Saved to: /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/embeddings/prompt_embeddings.pkl


================================================================

STEP 7 IMAGE EMBEDDING EXTRACTION (WITH PREPROCESSING)


    Extract CLIP embedding with comprehensive preprocessing.
    
    Pipeline:
    1. Validate image
    2. Preprocess (EXIF, RGB conversion, aspect ratio)
    3. CLIP processor (normalization, tensor conversion)
    4. Extract embedding
    5. L2 normalization
    
    FIXED: Properly handles CLIP output tensors


================================================================

In [11]:
def extract_prompts_for_image(image_embedding, theme, prompt_embeddings, top_k=2):
    """Extract top prompts using theme-first approach."""

    if theme not in prompt_embeddings:
        return {}

    theme_prompts = prompt_embeddings[theme]
    extracted_prompts = {}

    for category, prompts in theme_prompts.items():
        scores = []

        for prompt_data in prompts:
            score = np.dot(image_embedding, prompt_data['embedding'])
            scores.append((prompt_data['text'], float(score)))

        top_prompts = sorted(scores, key=lambda x: x[1], reverse=True)[:top_k]
        #filtered = [(text, score) for text, score in top_prompts if score > 0.40]
        filtered = [(text, score) for text, score in top_prompts if score > 0.25]  # Changed from 0.40

        if filtered:
            extracted_prompts[category] = [
                {'text': text, 'score': score} for text, score in filtered
            ]

    return extracted_prompts



def process_all_images(metadata, landmarks_path, model, processor, preprocessor, device):
    """Process all images with preprocessing."""

    image_embeddings = {}
    image_metadata_list = {}

    total_images = metadata['total_images']
    print(f"Processing {total_images} images...")

    stats = {'processed': 0, 'failed': 0}

    with tqdm(total=total_images, desc="Extracting embeddings") as pbar:
        for theme in metadata['themes']:
            theme_name = theme['theme_name']

            for state in theme['states']:
                state_name = state['state_name']

                for destination in state['destinations']:
                    dest_id = destination['destination_id']
                    dest_folder = destination['folder']

                    for img_filename in destination['images']:
                        image_id = f"{dest_id}_{img_filename.replace('.jpg', '').replace('.png', '')}"
                        image_path = os.path.join(landmarks_path, dest_folder, img_filename)

                        embedding = extract_image_embedding(image_path, model, processor, preprocessor, device)

                        if embedding is not None:
                            image_embeddings[image_id] = embedding

                            image_metadata_list[image_id] = {
                                'image_path': image_path,
                                'filename': img_filename,
                                'destination_id': dest_id,
                                'destination_name': destination['destination_name'],
                                'theme': theme_name,
                                'state': state_name,
                                'folder': dest_folder
                            }
                            stats['processed'] += 1
                        else:
                            stats['failed'] += 1

                        pbar.update(1)

    print(f"\nProcessed: {stats['processed']}")
    if stats['failed'] > 0:
        print(f"Failed: {stats['failed']}")

    return image_embeddings, image_metadata_list, stats


image_embeddings, image_metadata_dict, validation_stats = process_all_images(
    metadata, LANDMARKS_PATH, model, processor, preprocessor, device
)

print("="*80)


Processing 805 images...


Extracting embeddings: 100%|██████████| 805/805 [07:22<00:00,  1.82it/s]


Processed: 805


===========================================================================================================

STEP 8: Prompt Extraction For all images

===========================================================================================================

In [12]:
def process_all_prompts(image_embeddings, image_metadata_dict, prompt_embeddings):
    """Extract prompts for all images."""

    all_image_prompts = {}

    print(f"Extracting prompts for {len(image_embeddings)} images...")

    with tqdm(total=len(image_embeddings), desc="Extracting prompts") as pbar:
        for image_id, embedding in image_embeddings.items():
            metadata = image_metadata_dict[image_id]
            theme = metadata['theme']

            prompts = extract_prompts_for_image(embedding, theme, prompt_embeddings, top_k=2)

            all_image_prompts[image_id] = {
                'image_id': image_id,
                'theme': theme,
                'destination_id': metadata['destination_id'],
                'extracted_prompts': prompts,
                'total_prompts_extracted': sum(len(p) for p in prompts.values())
            }

            pbar.update(1)

    print(f"Extracted prompts for {len(all_image_prompts)} images")
    return all_image_prompts


all_image_prompts = process_all_prompts(image_embeddings, image_metadata_dict, prompt_embeddings)

print("="*80)


Extracting prompts for 805 images...


Extracting prompts: 100%|██████████| 805/805 [00:00<00:00, 997.80it/s] 

Extracted prompts for 805 images


===========================================================================================================

STEP 9: AGGREGATE PER DESTINATION

===========================================================================================================

In [13]:
print("\n" + "="*80)
print("STAGE 9: AGGREGATING BY DESTINATION")
print("="*80)

def aggregate_destination_embeddings(image_embeddings, image_metadata_dict):
    """Aggregate embeddings per destination."""

    destination_embeddings = {}
    destination_image_embeddings = {}

    for image_id, embedding in image_embeddings.items():
        dest_id = image_metadata_dict[image_id]['destination_id']

        if dest_id not in destination_image_embeddings:
            destination_image_embeddings[dest_id] = []

        destination_image_embeddings[dest_id].append(embedding)

    for dest_id, embeddings in destination_image_embeddings.items():
        embeddings_array = np.array(embeddings)
        avg_embedding = np.mean(embeddings_array, axis=0)
        avg_embedding = avg_embedding / np.linalg.norm(avg_embedding)

        destination_embeddings[dest_id] = {
            'average_embedding': avg_embedding,
            'individual_embeddings': embeddings_array,
            'num_images': len(embeddings)
        }

    print(f"Aggregated {len(destination_embeddings)} destinations")
    return destination_embeddings


def aggregate_destination_prompts(all_image_prompts):
    """Aggregate prompts with weighted scoring."""

    destination_prompts = {}
    dest_groups = {}

    for image_id, data in all_image_prompts.items():
        dest_id = data['destination_id']
        if dest_id not in dest_groups:
            dest_groups[dest_id] = []
        dest_groups[dest_id].append(data)

    for dest_id, images_data in dest_groups.items():
        num_images = len(images_data)
        category_prompts = {}

        for img_data in images_data:
            for category, prompts in img_data['extracted_prompts'].items():
                if category not in category_prompts:
                    category_prompts[category] = []
                for prompt in prompts:
                    category_prompts[category].append(prompt)

        aggregated = {}

        for category, prompts in category_prompts.items():
            prompt_stats = {}

            for prompt in prompts:
                text = prompt['text']
                score = prompt['score']

                if text not in prompt_stats:
                    prompt_stats[text] = {'scores': [], 'count': 0}

                prompt_stats[text]['scores'].append(score)
                prompt_stats[text]['count'] += 1

            weighted_prompts = []
            for text, stats in prompt_stats.items():
                avg_score = np.mean(stats['scores'])
                frequency = stats['count'] / num_images
                weighted_score = (avg_score * 0.6) + (frequency * 0.4)

                weighted_prompts.append({
                    'text': text,
                    'avg_score': float(avg_score),
                    'frequency': float(frequency),
                    'weighted_score': float(weighted_score),
                    'appearances': stats['count']
                })

            top_prompts = sorted(weighted_prompts, key=lambda x: x['weighted_score'], reverse=True)[:2]

            if top_prompts:
                aggregated[category] = top_prompts

        destination_prompts[dest_id] = {
            'destination_id': dest_id,
            'num_images': num_images,
            'aggregated_prompts': aggregated,
            'dominant_characteristics': {
                cat: prompts[0]['text'] for cat, prompts in aggregated.items() if prompts
            }
        }

    print(f"Aggregated prompts for {len(destination_prompts)} destinations")
    return destination_prompts


destination_embeddings = aggregate_destination_embeddings(image_embeddings, image_metadata_dict)
destination_prompts = aggregate_destination_prompts(all_image_prompts)

print("="*80)


STAGE 9: AGGREGATING BY DESTINATION
Aggregated 168 destinations
Aggregated prompts for 168 destinations


==========================================================================================================
STEP 10: SAVE ALL DATA

===========================================================================================================

In [14]:
print("\n" + "="*80)
print("STAGE 10: SAVING DATA")
print("="*80)

print("Saving embeddings...")

image_ids = list(image_embeddings.keys())
embeddings_array = np.array([image_embeddings[img_id] for img_id in image_ids])

destination_ids = list(destination_embeddings.keys())
dest_avg_embs = np.array([destination_embeddings[d_id]['average_embedding'] for d_id in destination_ids])

np.savez(
    f'{EMBEDDINGS_PATH}/all_embeddings.npz',
    image_ids=image_ids,
    image_embeddings=embeddings_array,
    destination_ids=destination_ids,
    destination_embeddings=dest_avg_embs
)

print(f"all_embeddings.npz ({embeddings_array.shape})")

embedding_index = {
    'image_index': {img_id: idx for idx, img_id in enumerate(image_ids)},
    'destination_index': {d_id: idx for idx, d_id in enumerate(destination_ids)},
    'metadata': {
        'created_date': datetime.now().isoformat(),
        'total_images': len(image_ids),
        'total_destinations': len(destination_ids),
        'embedding_dim': embeddings_array.shape[1],
        'model_name': model_name,
        'preprocessing_enabled': True,
        'validation_stats': validation_stats,
        'prompt_validation': validation_results
    }
}

with open(f'{EMBEDDINGS_PATH}/embedding_index.json', 'w') as f:
    json.dump(embedding_index, f, indent=2)

print(f"embedding_index.json")

with open(f'{PROMPTS_PATH}/image_prompts.json', 'w') as f:
    json.dump(all_image_prompts, f, indent=2)

print(f"image_prompts.json")

with open(f'{PROMPTS_PATH}/destination_prompts.json', 'w') as f:
    json.dump(destination_prompts, f, indent=2)

print(f"destination_prompts.json")

with open(f'{EMBEDDINGS_PATH}/destination_embeddings_detailed.pkl', 'wb') as f:
    pickle.dump(destination_embeddings, f)

print(f"destination_embeddings_detailed.pkl")

print("="*80)


STAGE 10: SAVING DATA
Saving embeddings...
all_embeddings.npz ((805, 512))
embedding_index.json
image_prompts.json
destination_prompts.json
destination_embeddings_detailed.pkl


==========================================================================================================
STEP 11: UPDATE METADATA

===========================================================================================================

In [15]:
print("\n" + "="*80)
print("STAGE 11: UPDATING METADATA")
print("="*80)

metadata['pipeline_status']['embeddings_computed'] = True
metadata['pipeline_status']['prompts_extracted'] = True
metadata['pipeline_status']['prompts_validated'] = True
metadata['last_updated'] = datetime.now().isoformat()
metadata['vl_encoding_version'] = '4.1'

for theme in metadata['themes']:
    for state in theme['states']:
        for destination in state['destinations']:
            dest_id = destination['destination_id']

            if dest_id in destination_embeddings:
                destination['embeddings_computed'] = True
                destination['prompts_extracted'] = True

                destination['embedding_references'] = {
                    'embeddings_file': 'vl_encoding/embeddings/all_embeddings.npz',
                    'destination_index': embedding_index['destination_index'][dest_id]
                }

                if dest_id in destination_prompts:
                    destination['dominant_prompts'] = destination_prompts[dest_id]['dominant_characteristics']

with open(METADATA_PATH, 'w') as f:
    json.dump(metadata, f, indent=2)

print(f"Updated metadata.json")
print("="*80)


STAGE 11: UPDATING METADATA
Updated metadata.json


==========================================================================================================

**STEP 12: VL Encoding and Verification Check**

Verify prompt extraction quality and alignment statistics

===========================================================================================================

In [16]:
import json
import numpy as np
from collections import Counter, defaultdict

print("="*80)
print("VL ENCODING VERIFICATION")
print("="*80)

# Paths
BASE_PATH = '/content/drive/MyDrive/visual-intelligence-travel-finance'
IMAGE_PROMPTS_PATH = f'{BASE_PATH}/data/vl_encoding/prompts/image_prompts.json'
DEST_PROMPTS_PATH = f'{BASE_PATH}/data/vl_encoding/prompts/destination_prompts.json'
METADATA_PATH = f'{BASE_PATH}/data/landmarks/metadata.json'

# Load data
print("\nLoading data...")
with open(IMAGE_PROMPTS_PATH, 'r') as f:
    image_prompts = json.load(f)

with open(DEST_PROMPTS_PATH, 'r') as f:
    destination_prompts = json.load(f)

with open(METADATA_PATH, 'r') as f:
    metadata = json.load(f)

print(f"Image prompts: {len(image_prompts)} images")
print(f"Destination prompts: {len(destination_prompts)} destinations")
print("="*80)


VL ENCODING VERIFICATION

Loading data...
Image prompts: 805 images
Destination prompts: 168 destinations


In [17]:
# VERIFICATION 1: OVERALL STATISTICS
print("\n" + "="*80)
print("1. OVERALL STATISTICS")
print("="*80)

total_images = len(image_prompts)
total_prompts_extracted = sum(data['total_prompts_extracted'] for data in image_prompts.values())
avg_prompts_per_image = total_prompts_extracted / total_images if total_images > 0 else 0

print(f"Total images processed: {total_images}")
print(f"Total prompts extracted: {total_prompts_extracted}")
print(f"Average prompts per image: {avg_prompts_per_image:.2f}")




1. OVERALL STATISTICS
Total images processed: 805
Total prompts extracted: 15401
Average prompts per image: 19.13


In [18]:
# VERIFICATION 2: CATEGORY COVERAGE
print("\n" + "="*80)
print("2. CATEGORY COVERAGE")
print("="*80)

all_categories = set()
category_counts = Counter()
images_per_category = defaultdict(int)

for image_id, data in image_prompts.items():
    for category in data['extracted_prompts'].keys():
        all_categories.add(category)
        category_counts[category] += len(data['extracted_prompts'][category])
        images_per_category[category] += 1

print(f"Total unique categories: {len(all_categories)}")
print(f"\nCategories found: {', '.join(sorted(all_categories))}")

print("\nPrompts per category:")
for category, count in category_counts.most_common():
    coverage = (images_per_category[category] / total_images) * 100
    print(f"  {category:20s}: {count:4d} prompts ({images_per_category[category]:3d} images, {coverage:5.1f}% coverage)")



2. CATEGORY COVERAGE
Total unique categories: 11

Categories found: Accessibility, Activities, Atmosphere, CrowdDensity, EconomyBudget, LandscapeType, NaturalVsCultural, RegionalStyle, VegetationType, VisualQuality, WaterFeatures

Prompts per category:
  Activities          : 1500 prompts (753 images,  93.5% coverage)
  Atmosphere          : 1461 prompts (739 images,  91.8% coverage)
  VisualQuality       : 1460 prompts (737 images,  91.6% coverage)
  CrowdDensity        : 1456 prompts (736 images,  91.4% coverage)
  NaturalVsCultural   : 1435 prompts (730 images,  90.7% coverage)
  Accessibility       : 1398 prompts (707 images,  87.8% coverage)
  WaterFeatures       : 1393 prompts (705 images,  87.6% coverage)
  LandscapeType       : 1389 prompts (707 images,  87.8% coverage)
  EconomyBudget       : 1376 prompts (695 images,  86.3% coverage)
  VegetationType      : 1278 prompts (659 images,  81.9% coverage)
  RegionalStyle       : 1255 prompts (637 images,  79.1% coverage)


In [19]:
# VERIFICATION 3: PROMPTS PER IMAGE DISTRIBUTION
print("\n" + "="*80)
print("3. PROMPTS PER IMAGE DISTRIBUTION")
print("="*80)

prompts_distribution = Counter()
for data in image_prompts.values():
    prompts_distribution[data['total_prompts_extracted']] += 1

print("Distribution of prompts extracted per image:")
for num_prompts in sorted(prompts_distribution.keys()):
    count = prompts_distribution[num_prompts]
    percentage = (count / total_images) * 100
    bar = "█" * int(percentage / 2)
    print(f"  {num_prompts:2d} prompts: {count:4d} images ({percentage:5.1f}%) {bar}")



3. PROMPTS PER IMAGE DISTRIBUTION
Distribution of prompts extracted per image:
   0 prompts:   19 images (  2.4%) █
   1 prompts:    4 images (  0.5%) 
   2 prompts:   10 images (  1.2%) 
   3 prompts:   12 images (  1.5%) 
   4 prompts:    2 images (  0.2%) 
   5 prompts:    7 images (  0.9%) 
   6 prompts:    3 images (  0.4%) 
   7 prompts:    5 images (  0.6%) 
   8 prompts:    7 images (  0.9%) 
   9 prompts:    3 images (  0.4%) 
  10 prompts:    7 images (  0.9%) 
  11 prompts:    5 images (  0.6%) 
  12 prompts:    4 images (  0.5%) 
  13 prompts:   13 images (  1.6%) 
  14 prompts:   12 images (  1.5%) 
  15 prompts:    8 images (  1.0%) 
  16 prompts:   10 images (  1.2%) 
  17 prompts:    9 images (  1.1%) 
  18 prompts:   27 images (  3.4%) █
  19 prompts:   14 images (  1.7%) 
  20 prompts:  109 images ( 13.5%) ██████
  21 prompts:   43 images (  5.3%) ██
  22 prompts:  472 images ( 58.6%) █████████████████████████████


In [20]:
# Find images with no prompts
no_prompts = [img_id for img_id, data in image_prompts.items() if data['total_prompts_extracted'] == 0]
if no_prompts:
    print(f"\nWARNING: {len(no_prompts)} images have NO prompts extracted!")
    print("Sample images with no prompts:")
    for img_id in no_prompts[:5]:
        print(f"  - {img_id}")



Sample images with no prompts:
  - HILLSTATION_HIMACHAL_PRADESH_CHAMBA_chamba_003
  - HILLSTATION_HIMACHAL_PRADESH_DALHOUSIE_dalhousie_001
  - HILLSTATION_HIMACHAL_PRADESH_DALHOUSIE_dalhousie_002
  - HILLSTATION_HIMACHAL_PRADESH_DALHOUSIE_dalhousie_003
  - HILLSTATION_HIMACHAL_PRADESH_DALHOUSIE_dalhousie_004


In [21]:
# VERIFICATION 4: SCORE DISTRIBUTION
print("\n" + "="*80)
print("4. SCORE DISTRIBUTION")
print("="*80)

all_scores = []
for data in image_prompts.values():
    for category, prompts in data['extracted_prompts'].items():
        for prompt in prompts:
            all_scores.append(prompt['score'])

if all_scores:
    print(f"Total prompt-image matches: {len(all_scores)}")
    print(f"Score statistics:")
    print(f"  Min:    {min(all_scores):.4f}")
    print(f"  Max:    {max(all_scores):.4f}")
    print(f"  Mean:   {np.mean(all_scores):.4f}")
    print(f"  Median: {np.median(all_scores):.4f}")
    print(f"  Std:    {np.std(all_scores):.4f}")

    # Score ranges
    print("\nScore ranges:")
    ranges = [
        (0.25, 0.30, 'Weak Match'),
        (0.30, 0.35, 'Decent Match'),
        (0.35, 1.00, 'Excellent match')
    ]

    for min_score, max_score, label in ranges:
        count = sum(1 for s in all_scores if min_score <= s < max_score)
        percentage = (count / len(all_scores)) * 100
        print(f"  {label:15s} ({min_score:.2f}-{max_score:.2f}): {count:4d} ({percentage:5.1f}%)")



4. SCORE DISTRIBUTION
Total prompt-image matches: 15401
Score statistics:
  Min:    0.2500
  Max:    0.3571
  Mean:   0.2903
  Median: 0.2897
  Std:    0.0213

Score ranges:
  Weak Match      (0.25-0.30): 10468 ( 68.0%)
  Decent Match    (0.30-0.35): 4886 ( 31.7%)
  Excellent match (0.35-1.00):   47 (  0.3%)


In [22]:
# VERIFICATION 5: THEME-SPECIFIC ANALYSIS
print("\n" + "="*80)
print("5. THEME-SPECIFIC ANALYSIS")
print("="*80)

theme_stats = defaultdict(lambda: {'images': 0, 'prompts': 0, 'categories': set()})

for image_id, data in image_prompts.items():
    theme = data['theme']
    theme_stats[theme]['images'] += 1
    theme_stats[theme]['prompts'] += data['total_prompts_extracted']
    theme_stats[theme]['categories'].update(data['extracted_prompts'].keys())

print("Statistics by theme:")
for theme, stats in sorted(theme_stats.items()):
    avg_prompts = stats['prompts'] / stats['images'] if stats['images'] > 0 else 0
    print(f"\n{theme}:")
    print(f"  Images: {stats['images']}")
    print(f"  Total prompts: {stats['prompts']}")
    print(f"  Avg prompts/image: {avg_prompts:.2f}")
    print(f"  Categories used: {len(stats['categories'])}")
    print(f"  Categories: {', '.join(sorted(stats['categories']))}")


5. THEME-SPECIFIC ANALYSIS
Statistics by theme:

Beach:
  Images: 229
  Total prompts: 4778
  Avg prompts/image: 20.86
  Categories used: 11
  Categories: Accessibility, Activities, Atmosphere, CrowdDensity, EconomyBudget, LandscapeType, NaturalVsCultural, RegionalStyle, VegetationType, VisualQuality, WaterFeatures

HillStation:
  Images: 240
  Total prompts: 3840
  Avg prompts/image: 16.00
  Categories used: 11
  Categories: Accessibility, Activities, Atmosphere, CrowdDensity, EconomyBudget, LandscapeType, NaturalVsCultural, RegionalStyle, VegetationType, VisualQuality, WaterFeatures

Temple:
  Images: 251
  Total prompts: 5105
  Avg prompts/image: 20.34
  Categories used: 11
  Categories: Accessibility, Activities, Atmosphere, CrowdDensity, EconomyBudget, LandscapeType, NaturalVsCultural, RegionalStyle, VegetationType, VisualQuality, WaterFeatures

Waterfall:
  Images: 85
  Total prompts: 1678
  Avg prompts/image: 19.74
  Categories used: 11
  Categories: Accessibility, Activities, 

In [23]:
# VERIFICATION 6: DESTINATION AGGREGATION QUALITY
print("\n" + "="*80)
print("6. DESTINATION AGGREGATION QUALITY")
print("="*80)

dest_with_prompts = sum(1 for d in destination_prompts.values()
                        if len(d.get('aggregated_prompts', {})) > 0)
total_destinations = len(destination_prompts)

print(f"Destinations with prompts: {dest_with_prompts}/{total_destinations}")

# Sample destination analysis
sample_dest_id = list(destination_prompts.keys())[0]
sample = destination_prompts[sample_dest_id]

print(f"\nSample destination: {sample_dest_id}")
print(f"Number of images: {sample['num_images']}")
print(f"Categories with prompts: {len(sample['aggregated_prompts'])}")

if sample['aggregated_prompts']:
    print("\nTop prompts per category:")
    for category, prompts in sorted(sample['aggregated_prompts'].items()):
        print(f"\n  {category}:")
        for prompt in prompts[:2]:
            print(f"    - {prompt['text']}")
            print(f"      (avg: {prompt['avg_score']:.3f}, freq: {prompt['frequency']:.2f}, weighted: {prompt['weighted_score']:.3f})")



6. DESTINATION AGGREGATION QUALITY
Destinations with prompts: 168/168

Sample destination: BEACH_GOA_AGONDA_BEACH
Number of images: 1
Categories with prompts: 10

Top prompts per category:

  Accessibility:
    - road-accessible beach, midday sunlight, dry season, empty surroundings, close-up framing
      (avg: 0.279, freq: 1.00, weighted: 0.567)
    - road-accessible beach, dramatic shadows, summer season, empty surroundings, close-up framing
      (avg: 0.277, freq: 1.00, weighted: 0.566)

  Activities:
    - beach walking and relaxation, diffused natural light, dry season, empty surroundings, close-up framing
      (avg: 0.285, freq: 1.00, weighted: 0.571)
    - beach walking and relaxation, dramatic shadows, dry season, sparse local presence, close-up framing
      (avg: 0.273, freq: 1.00, weighted: 0.564)

  Atmosphere:
    - warm humid seaside atmosphere, midday sunlight, summer season, empty surroundings, elevated viewpoint
      (avg: 0.265, freq: 1.00, weighted: 0.559)
    -

In [24]:
# VERIFICATION 7: MOST COMMON PROMPTS
print("\n" + "="*80)
print("7. MOST COMMON PROMPTS ACROSS ALL IMAGES")
print("="*80)

prompt_frequency = Counter()
for data in image_prompts.values():
    for category, prompts in data['extracted_prompts'].items():
        for prompt in prompts:
            prompt_frequency[prompt['text']] += 1

print("Top 20 most frequent prompts:")
for i, (prompt_text, count) in enumerate(prompt_frequency.most_common(20), 1):
    percentage = (count / total_images) * 100
    print(f"{i:2d}. {prompt_text:60s} ({count:4d} images, {percentage:5.1f}%)")



7. MOST COMMON PROMPTS ACROSS ALL IMAGES
Top 20 most frequent prompts:
 1. pilgrimage-based local economy, bright clear daylight, post-monsoon greenery, light visitor activity, wide-angle view (  97 images,  12.0%)
 2. coconut palm groves, golden hour glow, dry season, light visitor activity, ground-level view (  94 images,  11.7%)
 3. high-detail temple architecture view, low contrast light, monsoon season, light visitor activity, panoramic perspective (  87 images,  10.8%)
 4. high-detail temple architecture view, low contrast light, monsoon season, empty surroundings, panoramic perspective (  86 images,  10.7%)
 5. devotee presence, low contrast light, summer season, busy peak hours, panoramic perspective (  83 images,  10.3%)
 6. temple gardens and sacred trees, bright clear daylight, dry season, empty surroundings, panoramic perspective (  80 images,   9.9%)
 7. temple gardens and sacred trees, warm evening tones, monsoon season, light visitor activity, ground-level view (  79 im

In [25]:
# VERIFICATION 8: QUALITY CHECKS
print("\n" + "="*80)
print("8. QUALITY CHECKS")
print("="*80)

issues = []

# Check 1: Images with very few prompts
low_prompt_threshold = 5
low_prompt_images = sum(1 for data in image_prompts.values()
                        if data['total_prompts_extracted'] < low_prompt_threshold)
if low_prompt_images > 0:
    percentage = (low_prompt_images / total_images) * 100
    issues.append(f"{low_prompt_images} images ({percentage:.1f}%) have fewer than {low_prompt_threshold} prompts")

# Check 2: Categories with low coverage
low_coverage_threshold = 50  # percentage
for category, coverage_pct in [(cat, (images_per_category[cat] / total_images) * 100)
                               for cat in all_categories]:
    if coverage_pct < low_coverage_threshold:
        issues.append(f"Category '{category}' has low coverage: {coverage_pct:.1f}%")

# Check 3: Low average scores
if all_scores and np.mean(all_scores) < 0.50:
    issues.append(f"Average matching score is low: {np.mean(all_scores):.3f}")

# Check 4: Destinations without prompts
dest_no_prompts = sum(1 for d in destination_prompts.values()
                      if len(d.get('aggregated_prompts', {})) == 0)
if dest_no_prompts > 0:
    issues.append(f"{dest_no_prompts} destinations have no aggregated prompts")

if issues:
    print("WARNING - Issues found:")
    for issue in issues:
        print(f"  - {issue}")
else:
    print("All quality checks passed!")


8. QUALITY CHECKS
WARNING - Issues found:
  - 47 images (5.8%) have fewer than 5 prompts
  - Average matching score is low: 0.290


In [26]:
# VERIFICATION 9: ALIGNMENT VERIFICATION
print("\n" + "="*80)
print("9. PROMPT-IMAGE ALIGNMENT VERIFICATION")
print("="*80)

# Sample random images and show their top prompts
import random
sample_size = 5
sample_images = random.sample(list(image_prompts.keys()), min(sample_size, len(image_prompts)))

print(f"Random sample of {len(sample_images)} images with their extracted prompts:\n")

for img_id in sample_images:
    data = image_prompts[img_id]
    print(f"Image: {img_id}")
    print(f"Theme: {data['theme']}")
    print(f"Total prompts: {data['total_prompts_extracted']}")

    if data['extracted_prompts']:
        print("Top prompts by category:")
        for category, prompts in sorted(data['extracted_prompts'].items()):
            print(f"  {category}:")
            for prompt in prompts[:2]:
                print(f"    - {prompt['text']} (score: {prompt['score']:.3f})")
    print()




9. PROMPT-IMAGE ALIGNMENT VERIFICATION
Random sample of 5 images with their extracted prompts:

Image: TEMPLE_KARNATAKA_UDUPI_KRISHNA_TEMPLE_udupi_krishna_temple_002
Theme: Temple
Total prompts: 22
Top prompts by category:
  Accessibility:
    - pilgrimage-access temple, midday sunlight, dry season, sparse local presence, wide-angle view (score: 0.315)
    - pilgrimage-access temple, midday sunlight, monsoon season, empty surroundings, panoramic perspective (score: 0.314)
  Activities:
    - temple pilgrimage and worship, dramatic shadows, post-monsoon greenery, empty surroundings, wide-angle view (score: 0.318)
    - temple pilgrimage and worship, dramatic shadows, post-monsoon greenery, light visitor activity, ground-level view (score: 0.314)
  Atmosphere:
    - calm spiritual temple ambiance, early morning light, dry season, sparse local presence, panoramic perspective (score: 0.304)
    - calm spiritual temple ambiance, low contrast light, post-monsoon greenery, sparse local prese

In [27]:
# SUMMARY
print("="*80)
print("VERIFICATION SUMMARY")
print("="*80)
print(f"Total images analyzed: {total_images}")
print(f"Total prompts extracted: {total_prompts_extracted}")
print(f"Categories identified: {len(all_categories)}")
print(f"Average prompts per image: {avg_prompts_per_image:.2f}")
print(f"Average matching score: {np.mean(all_scores):.3f}" if all_scores else "No scores available")
print(f"Destinations with prompts: {dest_with_prompts}/{total_destinations}")

if issues:
    print(f"\nWARNING: {len(issues)} quality issues detected (see section 8)")
else:
    print("\nAll quality checks passed!")

print("="*80)

VERIFICATION SUMMARY
Total images analyzed: 805
Total prompts extracted: 15401
Categories identified: 11
Average prompts per image: 19.13
Average matching score: 0.290
Destinations with prompts: 168/168



**Visual Spot Check**


In [28]:

print("\n" + "="*80)
print("QUICK SPOT CHECK")
print("="*80)

# Pick random image
random_img_id = np.random.choice(list(image_embeddings.keys()))
print(f"\nRandom Image: {random_img_id}")
print(f"Destination: {image_metadata_dict[random_img_id]['destination_name']}")
print(f"Has embedding: {random_img_id in image_embeddings}")
print(f"Embedding shape: {image_embeddings[random_img_id].shape}")
print(f"Has prompts: {random_img_id in all_image_prompts}")
print(f"Prompt count: {all_image_prompts[random_img_id]['total_prompts_extracted']}")

print("\nExtracted prompts:")
for cat, prompts in all_image_prompts[random_img_id]['extracted_prompts'].items():
    print(f"\n  {cat}:")
    for p in prompts:
        print(f"    '{p['text']}' ({p['score']:.3f})")


QUICK SPOT CHECK

Random Image: TEMPLE_UTTAR_PRADESH_HANUMAN_TEMPLE_ALLAHABAD_hanuman_temple_allahabad_001
Destination: Hanuman Temple Allahabad
Has embedding: True
Embedding shape: (512,)
Has prompts: True
Prompt count: 3

Extracted prompts:

  Atmosphere:
    'calm spiritual temple ambiance, warm evening tones, post-monsoon greenery, light visitor activity, wide-angle view' (0.253)
    'calm spiritual temple ambiance, warm evening tones, monsoon season, empty surroundings, elevated viewpoint' (0.252)

  CrowdDensity:
    'devotee presence, warm evening tones, summer season, empty surroundings, wide-angle view' (0.253)


========================================================================================================== STEP 12: FINAL SUMMARY

===========================================================================================================

In [29]:
print("\n" + "="*80)
print("SUCCESS! VL ENCODING COMPLETE")
print("="*80)

print(f"\nImages processed: {len(image_embeddings)}")
print(f"Destinations: {len(destination_embeddings)}")
print(f"Embedding dimension: {embeddings_array.shape[1]}")
print(f"Model: {model_name}")

print(f"\nPrompt Validation:")
print(f"  Total prompts: {validation_results['total_prompts']}")
print(f"  Valid prompts: {validation_results['valid_prompts']}")
print(f"  Issues found: {validation_results['prompts_with_issues']}")

print("\nOutput Files:")
print(f"  {EMBEDDINGS_PATH}/all_embeddings.npz")
print(f"  {EMBEDDINGS_PATH}/embedding_index.json")
print(f"  {EMBEDDINGS_PATH}/destination_embeddings_detailed.pkl")
print(f"  {PROMPTS_PATH}/image_prompts.json")
print(f"  {PROMPTS_PATH}/destination_prompts.json")
print(f"  {REPORTS_PATH}/prompt_validation_report.txt")

print("\nCategory-Based Prompts Available:")
categories = set()
for img_prompts in all_image_prompts.values():
    categories.update(img_prompts['extracted_prompts'].keys())
print(f"  Categories extracted: {', '.join(sorted(categories))}")

print("\nReady for Category-Aware Matching!")
print("="*80)


SUCCESS! VL ENCODING COMPLETE

Images processed: 805
Destinations: 168
Embedding dimension: 512
Model: openai/clip-vit-base-patch32

Prompt Validation:
  Total prompts: 2200
  Valid prompts: 2200
  Issues found: 0

Output Files:
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/embeddings/all_embeddings.npz
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/embeddings/embedding_index.json
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/embeddings/destination_embeddings_detailed.pkl
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/prompts/image_prompts.json
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/prompts/destination_prompts.json
  /content/drive/MyDrive/visual-intelligence-travel-finance/data/vl_encoding/reports/prompt_validation_report.txt

Category-Based Prompts Available:
  Categories extracted: Accessibility, Activities, Atmosphere, Crow

#VL Encoding Layer - System Performance Report

**Overview**
Vision-Language Model (CLIP) based semantic encoding system for Indian tourism destinations. Extracts multi-attribute descriptions from travel images using zero-shot learning.

**Dataset Statistics**
<table>
<tr><th>Metric</th><th>Value</th></tr>
<tr><td>Total Images</td><td>805</td></tr>
<tr><td>Destinations</td><td>168</td></tr>
<tr><td>Prompts Extracted</td><td>15,401</td></tr>
<tr><td>Avg Prompts/Image</td><td>19.13</td></tr>
<tr><td>Embedding Dimension</td><td>512</td></tr>
</table>




**Theme Distribution**
<table>
<tr><th>Theme</th><th>Images</th><th>Prompts</th><th>Avg/Image</th></tr>
<tr><td>Beach</td><td>229 (28.4%)</td><td>4,778</td><td>20.86</td></tr>
<tr><td>Hill Station</td><td>240 (29.8%)</td><td>3,840</td><td>16.00</td></tr>
<tr><td>Temple</td><td>251 (31.2%)</td><td>5,105</td><td>20.34</td></tr>
<tr><td>Waterfall</td><td>85 (10.6%)</td><td>1,678</td><td>19.74</td></tr>
</table>




**Category Coverage**
11 semantic categories with 79-94% image coverage:
<table>
<tr><th>Category</th><th>Coverage</th><th>Prompts</th></tr>
<tr><td>Activities</td><td>93.5%</td><td>1,500</td></tr>
<tr><td>Atmosphere</td><td>91.8%</td><td>1,461</td></tr>
<tr><td>VisualQuality</td><td>91.6%</td><td>1,460</td></tr>
<tr><td>CrowdDensity</td><td>91.4%</td><td>1,456</td></tr>
<tr><td>NaturalVsCultural</td><td>90.7%</td><td>1,435</td></tr>
<tr><td>Accessibility</td><td>87.8%</td><td>1,398</td></tr>
<tr><td>WaterFeatures</td><td>87.6%</td><td>1,393</td></tr>
<tr><td>LandscapeType</td><td>87.8%</td><td>1,389</td></tr>
<tr><td>EconomyBudget</td><td>86.3%</td><td>1,376</td></tr>
<tr><td>VegetationType</td><td>81.9%</td><td>1,278</td></tr>
<tr><td>RegionalStyle</td><td>79.1%</td><td>1,255</td></tr>
</table>


**Similarity Scores**
Mean Score: 0.290 (appropriate for compositional prompts)
<table>
<tr><th>Score Range</th><th>Distribution</th><th>Label</th></tr>
<tr><td>0.35-1.00</td><td>0.3%</td><td>Excellent</td></tr>
<tr><td>0.30-0.35</td><td>31.7%</td><td>Decent</td></tr>
<tr><td>0.25-0.30</td><td>68.0%</td><td>Valid</td></tr>
</table>
Note: Scores of 0.25-0.35 are expected for multi-attribute prompts like "sandy coastal landscape, diffused natural light, dry season, moderate crowd" vs. simple prompts like "beach" (0.40+).


**Extraction Quality**


**Strengths**

100% destination coverage (168/168)
Consistent extraction: 58.6% images achieve full 22 prompts
Balanced theme coverage across all 4 types
All 11 categories actively used

**Issues Identified**

2.4% failure rate (19 images, primarily Dalhousie/Chamba hill stations)
5.8% images with less than 5 prompts
Hill stations lowest average (16.0 prompts/image)

**Top Semantic Patterns**
Most frequent prompts across dataset:


**Pilgrimage-based**

local economy (12.0% of images)
Coconut palm groves (11.7%)
High-detail temple architecture (10.8%)
Devotee presence (10.3%)
Temple gardens and sacred trees (9.9%)

**Use Cases Enabled**

Visual Similarity Search - Upload image and find similar destinations

Semantic Text Queries - "peaceful beach with low crowd" returns matched results

Multi-Category Filtering - Combine attributes (Activities + Atmosphere + Budget)

Destination Recommendations - "Similar to Agonda Beach" returns weighted matches

Accessibility Planning - Filter by mobility requirements

Budget-Aware Search - EconomyBudget category filtering


Processing: Zero-shot (no fine-tuning required)

Scalability: Linear O(n) with image count

Storage: Efficient npz format for embeddings

Query Speed: Real-time cosine similarity search


**Prompt Distribution Analysis**

Prompts per Image Distribution

22 prompts: 472 images (58.6%)

20 prompts: 109 images (13.5%)

18 prompts: 27 images (3.4%)

0-17 prompts: 197 images (24.5%)




**Category-wise Statistics**


Top performing categories by coverage:

Activities: 753 images (93.5%)

Atmosphere: 739 images (91.8%)

VisualQuality: 737 images (91.6%)


**Lower performing categories:**

VegetationType: 659 images (81.9%)

RegionalStyle: 637 images (79.1%)


#Conclusion

Production-ready semantic search system with 15,401 prompts across 168 destinations achieving 100% coverage. The multi-category taxonomy enables sophisticated filtering and personalized recommendations without manual tagging. System successfully bridges visual and textual modalities for intelligent travel planning.


Output Files Generated
/data/vl_encoding/embeddings/all_embeddings.npz
/data/vl_encoding/embeddings/embedding_index.json
/data/vl_encoding/embeddings/destination_embeddings_detailed.pkl
/data/vl_encoding/prompts/image_prompts.json
/data/vl_encoding/prompts/destination_prompts.json
/data/vl_encoding/reports/prompt_validation_report.txt